### PR TITLE
chore: refactor Views to extend a shared class

### PR DESCRIPTION
**Pull request by GoodBoyDigital**
##### Description of change

Added a new `ViewContainer` Class that contains most of the shared logic for Views. Reducing the amount of code and generally making things a bit nicer!

image

I also simplified how didViewUpdate works - this is now also used by the pipes rather the views (rather than an additonal prop like `didSpriteUpdate`)

A view now just needs to override `updateBounds` as the `boundsDirty` is handled by the shared class.

Tests have been updated too!

Not Happy with `ViewContainer` as the name though.. suggestions welcome :D

##### Pre-Merge Checklist

- [x] Tests and/or benchmarks are included
- [x] Documentation is changed or added
- [x] Lint process passed (`npm run lint`)
- [x] Tests passed (`npm run test`)

### CODE REVIEW

File: src/scene/container/ViewContainer.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+{
+    public readonly canBundle = true;
+    public renderPipeId: string;
+    public allowChildren = false;
+
+    public _roundPixels: 0 | 1 = 0;
+    public _lastUsed = 0;
+    public _lastInstructionTick = -1;
+    public _didUpdateTickCompare = -1;
+    public _didViewUpdate = false;
```
Comment (dev7355608) [inline src/scene/container/ViewContainer.ts:17]: These seem to be unused.

File: src/scene/container/Container.ts [hunk 2]
```diff
@@ -648,7 +648,6 @@ export class Container<C extends ContainerChild = ContainerChild> extends EventE
         child.parent = this;
 
         child.didChange = true;
-        child.didViewUpdate = false;
```
Comment (dev7355608) [inline src/scene/container/Container.ts:651]: `addChildAt` does still assign `child.didViewUpdate = false;`.

File: src/scene/container/Container.ts [hunk 3]
```diff
@@ -377,7 +377,7 @@ export class Container<C extends ContainerChild = ContainerChild> extends EventE
     public didChange = false;
     // same as above, but for the renderable
     /** @private */
-    public didViewUpdate = false;
+    public didViewUpdate = true;
```
Comment (dev7355608) [inline src/scene/container/Container.ts:380]: Do we still need this to be a property of `Container`. Shouldn't this be only a property on `ViewContainer`?

File: src/scene/sprite/Sprite.ts [hunk 4]
```diff
@@ -172,101 +157,11 @@ export class Sprite extends Container implements View
      * The local bounds of the sprite.
      * @type {rendering.Bounds}
      */
-    get bounds()
```
Comment (dev7355608) [inline src/scene/sprite/Sprite.ts:175]: These changes break rendering of sprites with trimmed textures, I assume.

File: src/scene/container/ViewContainer.ts (hunk 1 above)
Comment (GoodBoyDigital) [reply src/scene/container/ViewContainer.ts:17]: nice - removed!

File: src/scene/container/Container.ts (hunk 2 above)
Comment (GoodBoyDigital) [reply src/scene/container/Container.ts:651]: removed 👍

File: src/scene/sprite/Sprite.ts (hunk 4 above)
Comment (GoodBoyDigital) [reply src/scene/sprite/Sprite.ts:175]: this should be all good! i merged it so that bounds always returns the trimmed bounds if they exist.
This keeps things the same as they were. But will have a think about how to distinguish between the two. Something like `bounds` for when users measure and perhaps `viewBounds` which factors in trimmed textures which we can use internally (eg for filters measuring)

Comment (dev7355608) [reply src/scene/sprite/Sprite.ts:175]: The sprite should be rendered based on the `texture.trim` bounds (which doesn't appear to be the case in this PR), but the `bounds` of the sprite itself should not use `texture.trim` so that is has the same bounds like the same texture would have if it was untrimmed (this is correct in this PR). If we had a visual test for a trimmed texture, I'd expect it to fail and render stretched with this PR.

Comment (GoodBoyDigital) [reply src/scene/sprite/Sprite.ts:175]: let me put up a visual test for this in the PR

Comment (GoodBoyDigital) [reply src/scene/sprite/Sprite.ts:175]: test added - and trim code is back in. Thanks for clarifying!

**Comment by GoodBoyDigital** [comment]
Hey @dev7355608 , this should be good to go now! the bounds for sprites have also been corrected. Thanks for reviewing!

File: src/scene/container/RenderGroup.ts [hunk 5]
```diff
@@ -200,18 +201,7 @@ export class RenderGroup implements Instruction
         childrenToUpdate.list[childrenToUpdate.index++] = child;
     }
 
-    // SHOULD THIS BE HERE?
-    public updateRenderable(container: Container)
```
Comment (Zyie) [inline src/scene/container/RenderGroup.ts:204]: this seems like another public API being removed

File: src/scene/container/RenderContainer.ts [hunk 6]
```diff
@@ -106,9 +77,6 @@ export class RenderContainer extends Container implements View, Instruction
         });
 
         if (render) this.render = render;
-
-        this.containsPoint = options.containsPoint ?? (() => false);
-        this.addBounds = options.addBounds ?? (() => false);
```
Comment (Zyie) [inline src/scene/container/RenderContainer.ts:111]: why have these been removed?

File: src/rendering/renderers/shared/view/View.ts [hunk 7]
```diff
@@ -43,8 +42,6 @@ export interface View
     /** this is the AABB rectangle bounds of the view in local untransformed space. */
     bounds: BoundsData;
 
-    /** Adds the current bounds of this view to the supplied bounds */
-    addBounds: (bounds: Bounds) => void;
```
Comment (Zyie) [inline src/rendering/renderers/shared/view/View.ts:47]: `addBounds` is a public API so we shouldn't be removing it

File: src/scene/text/canvas/CanvasTextPipe.ts [hunk 8]
```diff
@@ -182,7 +180,7 @@ export class CanvasTextPipe implements RenderPipe<Text>
         this._gpuText[text.uid] = gpuTextData;
 
         text._resolution = text._autoResolution ? this._renderer.resolution : text.resolution;
-        this._updateText(text);
+        // this._updateText(text);
```
[the remark this answers is not in the document]
Comment (GoodBoyDigital) [reply src/scene/text/canvas/CanvasTextPipe.ts:183]: no longer needed - will remove

File: src/scene/sprite/SpritePipe.ts [hunk 9]
```diff
@@ -34,7 +34,9 @@ export class SpritePipe implements RenderPipe<Sprite>
     {
         const gpuSprite = this._getGpuSprite(sprite);
 
-        if (sprite._didSpriteUpdate) this._updateBatchableSprite(sprite, gpuSprite);
+        // console.log(sprite.didViewUpdate);
```
[the remark this answers is not in the document]
Comment (GoodBoyDigital) [reply src/scene/sprite/SpritePipe.ts:37]: agree!

File: src/scene/sprite/Sprite.ts [hunk 10]
```diff
@@ -172,100 +172,24 @@ export class Sprite extends Container implements View
...
-    {
-        if (this._sourceBoundsDirty)
-        {
-            this._updateSourceBounds();
-            this._sourceBoundsDirty = false;
-        }
-
-        return this._sourceBounds;
-    }
+    // public containsPoint(point: PointData): boolean
```
[the remark this answers is not in the document]
Comment (GoodBoyDigital) [reply src/scene/sprite/Sprite.ts:180]: agreed!

File: src/scene/sprite/Sprite.ts [hunk 11]
```diff
@@ -172,100 +172,24 @@ export class Sprite extends Container implements View
...
-
-        return this._bounds;
+        updateQuadBounds(this._bounds, this._anchor, this._texture, 0, false);
     }
 
-    /**
-     * The bounds of the sprite, taking the texture's trim into account.
-     * @type {rendering.Bounds}
-     */
-    get sourceBounds()
```
[the remark this answers is not in the document]
Comment (GoodBoyDigital) [reply src/scene/sprite/Sprite.ts:190]: will add deprecation to this one!

File: src/scene/sprite-nine-slice/NineSliceSprite.ts [hunk 12]
```diff
@@ -70,34 +67,24 @@ export interface NineSliceSpriteOptions extends ContainerOptions
...
     public static defaultOptions: NineSliceSpriteOptions = {
         /** @default Texture.EMPTY */
         texture: Texture.EMPTY,
     };
 
-    public _roundPixels: 0 | 1 = 0;
     public readonly renderPipeId = 'nineSliceSprite';
     public _texture: Texture;
 
-    public batched = true;
```
[the remark this answers is not in the document]
Comment (GoodBoyDigital) [reply src/scene/sprite-nine-slice/NineSliceSprite.ts:85]: not required 👍

File: src/scene/container/RenderGroup.ts (hunk 5 above)
Comment (GoodBoyDigital) [reply src/scene/container/RenderGroup.ts:204]: this one is kind of internal public iyswim? Not really meant to be used by others. Think its still worth depricating?

File: src/scene/container/RenderContainer.ts (hunk 6 above)
Comment (GoodBoyDigital) [reply src/scene/container/RenderContainer.ts:111]: we can rely on the default functions in ViewContainer instead

File: src/rendering/renderers/shared/view/View.ts (hunk 7 above)
Comment (GoodBoyDigital) [reply src/rendering/renderers/shared/view/View.ts:47]: I think i can deprecate this one.

File: src/scene/container/RenderContainer.ts (hunk 6 above)
Comment (Zyie) [reply src/scene/container/RenderContainer.ts:111]: but this removes the ability to set these functions through the constructor, so would be a breaking change?

Comment (GoodBoyDigital) [reply src/scene/container/RenderContainer.ts:111]: true! have added in some backwrds compatability!

File: src/scene/sprite/Sprite.ts [hunk 13]
```diff
@@ -409,4 +304,18 @@ export class Sprite extends Container implements View
...
+        super.onViewUpdate();
+    }
+
+    /**
+     * @deprecated since 8.2.5 (Use `renderBounds` instead)
+     */
+    get sourceBounds()
+    {
+        return this.renderBounds;
+    }
```
Comment (dev7355608) [inline src/scene/sprite/Sprite.ts:320]:
```suggestion
    /**
     * @deprecated since 8.2.5 (Use `bounds` instead)
     */
    get sourceBounds()
    {
        return this.bounds;
    }
```

File: src/scene/container/RenderContainer.ts [hunk 14]
```diff
@@ -19,8 +19,21 @@ export interface RenderContainerOptions extends ContainerOptions
     render?: RenderFunction;
     /** how to know if the custom render logic contains a point or not, used for interaction */
     containsPoint?: (point: Point) => boolean;
-    /** how to add the bounds of this object when measuring */
+    /** @deprecated how to add the bounds of this object when measuring */
     addBounds?: (bounds: BoundsData) => void;
```
Comment (Zyie) [inline src/scene/container/RenderContainer.ts:23]:
```suggestion
    /**
      * how to add the bounds of this object when measuring 
      * @deprecated since 8.3.0
      * @see ViewContainer#updateBounds
      */
    addBounds?: (bounds: BoundsData) => void;
```

File: src/scene/container/RenderContainer.ts [hunk 15]
```diff
@@ -56,37 +69,9 @@ export interface RenderContainerOptions extends ContainerOptions
...
-    public containsPoint: (point: Point) => boolean;
-    /**
-     * Adds the bounds of this text to the bounds object.
-     * @param bounds - The output bounds object.
-     */
-    public addBounds: (bounds: Bounds) => void;
-
-    public canBundle = false;
-    public readonly renderPipeId: string = 'customRender';
+    public renderPipeId = 'customRender';
```
Comment (Zyie) [inline src/scene/container/RenderContainer.ts:74]:
```suggestion
    public readonly renderPipeId: string = 'customRender';
```

File: src/scene/container/RenderContainer.ts [hunk 16]
```diff
@@ -107,8 +92,22 @@ export class RenderContainer extends Container implements View, Instruction
...
+        if (options.containsPoint)
+        {
+            this.containsPoint = options.containsPoint;
+        }
+
+        if (options.addBounds)
+        {
+            this.addBounds = options.addBounds;
+
+            deprecation('8.2.5', 'RenderContainer#addBounds no longer used, use RenderContainer#updateBounds instead');
```
Comment (Zyie) [inline src/scene/container/RenderContainer.ts:104]:
```suggestion
            deprecation('8.3.0', 'RenderContainer#addBounds no longer used, use RenderContainer#updateBounds instead');
```

File: src/scene/container/RenderGroup.ts [hunk 17]
```diff
@@ -267,6 +258,21 @@ export class RenderGroup implements Instruction
...
+     */
+    public updateRenderable(viewContainer: ViewContainer)
+    {
+        // only update if its visible!
+        if (viewContainer.globalDisplayStatus < 0b111) return;
+
+        viewContainer.didViewUpdate = false;
+        // actually updates the renderable..
+        this.instructionSet.renderPipes[viewContainer.renderPipeId].updateRenderable(viewContainer);
+    }
```
Comment (Zyie) [inline src/scene/container/RenderGroup.ts:274]: what is the purpose of no longer calling this function?
seems like a needless breaking change and caused duplication of code between `RenderGroupSystem` and `updateRenderGroupTransform`

File: src/scene/container/ViewContainer.ts [hunk 18]
```diff
@@ -0,0 +1,97 @@
+import { Bounds } from './bounds/Bounds';
+import { Container } from './Container';
+
+import type { PointData } from '../../maths/point/PointData';
+import type { View } from '../../rendering/renderers/shared/view/View';
+import type { BoundsData } from './bounds/Bounds';
+
+export class ViewContainer extends Container implements View
```
Comment (Zyie) [inline src/scene/container/ViewContainer.ts:8]: can you document this class please!

Comment (Zyie) [inline src/scene/container/ViewContainer.ts:8]: this class should be `abstract` since we wouldn't want someone to make one of these right?
it also means you can mark `addBounds` and `updateBounds` as abstract

File: src/scene/container/bounds/getGlobalBounds.ts [hunk 19]
```diff
@@ -84,12 +84,11 @@ export function _getGlobalBounds(
     }
     else
     {
-        if ((target as Renderable).addBounds)
+        if ((target as Renderable).bounds)
```
Comment (Zyie) [inline src/scene/container/bounds/getGlobalBounds.ts:87]: I think it may be better to do this change in a separate PR to reduce the amount of noise here and ensure we don't have any breaking changes

File: src/scene/mesh/shared/Mesh.ts [hunk 20]
```diff
@@ -252,21 +234,12 @@ export class Mesh<
...
 
     /**
-     * The local bounds of the mesh.
+     * Update local bounds of the mesh.
      * @type {rendering.Bounds}
      */
-    get bounds()
-    {
-        return this._geometry.bounds;
-    }
```
Comment (Zyie) [inline src/scene/mesh/shared/Mesh.ts:261]: why are we no longer returning the geometry bounds?

File: src/scene/sprite-nine-slice/NineSliceSprite.ts [hunk 21]
```diff
@@ -70,34 +67,24 @@ export interface NineSliceSpriteOptions extends ContainerOptions
...
 {
     /** The default options, used to override the initial values of any options passed in the constructor. */
     public static defaultOptions: NineSliceSpriteOptions = {
         /** @default Texture.EMPTY */
         texture: Texture.EMPTY,
     };
 
-    public _roundPixels: 0 | 1 = 0;
-    public readonly renderPipeId: string = 'nineSliceSprite';
+    public readonly renderPipeId = 'nineSliceSprite';
```
Comment (Zyie) [inline src/scene/sprite-nine-slice/NineSliceSprite.ts:78]:
```suggestion
    public readonly renderPipeId: string = 'nineSliceSprite';
```

File: src/scene/sprite-tiling/TilingSprite.ts [hunk 22]
```diff
@@ -130,24 +128,14 @@ export class TilingSprite extends Container implements View, Instruction
         applyAnchorToTexture: false,
     };
 
-    public readonly renderPipeId: string = 'tilingSprite';
-    public readonly canBundle = true;
-    public readonly batched = true;
+    public readonly renderPipeId = 'tilingSprite';
```
Comment (Zyie) [inline src/scene/sprite-tiling/TilingSprite.ts:131]:
```suggestion
    public readonly renderPipeId: string = 'tilingSprite';
```

File: src/scene/sprite/Sprite.ts [hunk 23]
```diff
@@ -128,14 +118,24 @@ export class Sprite extends Container implements View
...
 
-        this.allowChildren = false;
         this.roundPixels = roundPixels ?? false;
 
         // needs to be set after the container has initiated
         if (width !== undefined) this.width = width;
         if (height !== undefined) this.height = height;
     }
 
+    get renderBounds(): BoundsData
```
Comment (Zyie) [inline src/scene/sprite/Sprite.ts:128]: can you document this

File: src/scene/sprite/Sprite.ts [hunk 24]
```diff
@@ -409,4 +304,18 @@ export class Sprite extends Container implements View
...
     }
+
+    public override onViewUpdate()
+    {
+        this._renderBoundsDirty = true;
+        super.onViewUpdate();
+    }
+
+    /**
+     * @deprecated since 8.2.5 (Use `bounds` instead)
```
Comment (Zyie) [inline src/scene/sprite/Sprite.ts:315]:
```suggestion
     * @deprecated since 8.3.0 (Use `bounds` instead)
```

File: tests/renderering/scene/RenderGroupSystem.test.ts [hunk 25]
```diff
@@ -19,7 +19,7 @@ describe('RenderGroupSystem', () =>
 
         text.text = 'hello world 2';
 
-        expect(container.renderGroup.childrenRenderablesToUpdate.index).toEqual(1);
+        expect(container.renderGroup.childrenRenderablesToUpdate.index).toEqual(0);
```
Comment (Zyie) [inline tests/renderering/scene/RenderGroupSystem.test.ts:22]: what caused this change?

File: tests/renderering/text/Text.tests.ts [hunk 26]
```diff
@@ -51,6 +51,8 @@ describe('Text', () =>
 
             const renderer = await getWebGLRenderer({ resolution: 2 });
 
+            renderer.render({ container: text });
```
Comment (Zyie) [inline tests/renderering/text/Text.tests.ts:54]: what caused this change?
